### PR TITLE
Overall changes

### PR DESCRIPTION
**Pull request by 3Hren**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -134,10 +123,26 @@ TARGET_LINK_LIBRARIES(cocaine-runtime
     ${Boost_LIBRARIES}
     cocaine-core)
 
-SET_TARGET_PROPERTIES(cocaine-core cocaine-runtime PROPERTIES
-    COMPILE_FLAGS "-std=c++0x -W -Wall -Werror -pedantic")
+# Try and enable C++11.
+add_cxx_compiler_flag(-std=c++11)
+IF(NOT HAVE_CXX_FLAG_STD_CXX11)
```
Comment (kobolog) [inline CMakeLists.txt]: Shouldn't it be IF/ELSE?

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,23 +2,26 @@ CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
...
+SET(COCAINE_VERSION_RELEASE 3)
 
 SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
 
-IF(NOT APPLE)
-    OPTION(COCAINE_ALLOW_CGROUPS "Build CGroups support for Process Isolate" ON)
-ELSE()
-    OPTION(COCAINE_ALLOW_CGROUPS "Build CGroups support for Process Isolate" OFF)
+IF(APPLE)
+    SET(CMAKE_MACOSX_RPATH ON)
```
Comment (kobolog) [inline CMakeLists.txt]: Duplicate.

File: debian/cocaine-v12-runtime.init [hunk 3]
```diff
@@ -14,7 +14,7 @@
 # Author: Andrey Sibiryov <kobolog@yandex-team.ru>
 
 PATH=/sbin:/bin:/usr/sbin:/usr/bin
-NAME=cocaine-runtime
+NAME=cocaine-v12-runtime
```
Comment (kobolog) [inline debian/cocaine-v12-runtime.init]: Why the name change?

File: include/cocaine/detail/engine.hpp [hunk 4]
```diff
@@ -59,8 +62,13 @@ class execution_unit_t {
...
 
-    std::shared_ptr<session_t>
-    attach(const std::shared_ptr<asio::ip::tcp::socket>& ptr, const io::dispatch_ptr_t& dispatch);
+    template<class Socket>
+    std::shared_ptr<session<typename Socket::protocol_type>>
+    attach(std::unique_ptr<Socket> ptr, const io::dispatch_ptr_t& dispatch);
+
+    template<class Socket>
+    std::shared_ptr<session<typename Socket::protocol_type>>
+    attach(const std::shared_ptr<Socket>& ptr, const io::dispatch_ptr_t& dispatch);
```
Comment (kobolog) [inline include/cocaine/detail/engine.hpp]: Breaks ownership semantics for engines. Why?

File: include/cocaine/detail/service/logging.hpp [hunk 5]
```diff
@@ -43,8 +44,10 @@ class logging_t:
 
 private:
     void
-    on_emit(logging::priorities level, std::string&& source, std::string&& message,
-            blackhole::attribute::set_t&& attributes);
+    on_emit(logging::priorities level,
+            const std::string& source,
```
Comment (kobolog) [inline include/cocaine/detail/service/logging.hpp]: Why did you remove rvalue-references?

File: include/cocaine/rpc/asio/readable_stream.hpp [hunk 6]
```diff
@@ -95,6 +96,7 @@ class readable_stream:
             m_ring.resize(m_ring.size() * 2);
         }
 
+        namespace ph = std::placeholders;
```
Comment (kobolog) [inline include/cocaine/rpc/asio/readable_stream.hpp]: Why this thing was moved here?

File: include/cocaine/rpc/session.hpp [hunk 7]
```diff
@@ -89,8 +98,21 @@ class session_t:
     void
     pull();
 
+    // TODO: Do something with it.
+    template<class Event, class... Args>
     void
-    push(io::encoder_t::message_type&& message);
+    push(uint64_t channel_id, Args&&... args) {
```
Comment (kobolog) [inline include/cocaine/rpc/session.hpp]: Why is this templated now?

File: CMakeLists.txt (hunk 1 above)
Comment (3Hren) [inline CMakeLists.txt]: It writes:

```
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Failed
-- Performing Test HAVE_CXX_FLAG_STD_CXX0X
-- Performing Test HAVE_CXX_FLAG_STD_CXX0X - Success
```
which looks better for my opinion.

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (3Hren) [inline debian/cocaine-v12-runtime.init]: @antmat welcome!

File: include/cocaine/detail/engine.hpp (hunk 4 above)
Comment (3Hren) [inline include/cocaine/detail/engine.hpp]: Because of https://github.com/cocaine/cocaine-core/blob/v0.12/src/service/locator.cpp#L401

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (noxiouz) [inline debian/cocaine-v12-runtime.init]: To have the same name which the package have. Otherwise, the init script would not find ```defaults``` and binary.

File: include/cocaine/detail/service/logging.hpp (hunk 5 above)
Comment (3Hren) [inline include/cocaine/detail/service/logging.hpp]: Accidentally. Fixe.

File: CMakeLists.txt (hunk 2 above)
Comment (3Hren) [inline CMakeLists.txt]: Thanks. Fixed.

File: include/cocaine/rpc/asio/readable_stream.hpp (hunk 6 above)
Comment (3Hren) [inline include/cocaine/rpc/asio/readable_stream.hpp]: @antmat welcome!

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (kobolog) [inline debian/cocaine-v12-runtime.init]: Okay, why the name change for everything else?

File: include/cocaine/detail/engine.hpp (hunk 4 above)
Comment (kobolog) [inline include/cocaine/detail/engine.hpp]: Isn't it implicitly move-constructible?

Comment (3Hren) [inline include/cocaine/detail/engine.hpp]: C++ can't deduce properly dependent type name.

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (noxiouz) [inline debian/cocaine-v12-runtime.init]: It's a long story about conductor, repos and all of that. May be you remember the conductor? It can't handle a ticket (even create) if a package is not in all repos. So we had cocaine-runtime < 12 in common-precise repo and cocaine-runtime > 12 in our private repo cocaine-v12-precise. It makes impossible to use conductor to install the package. 
Long story short it's a Yandex related problem.

File: include/cocaine/detail/engine.hpp (hunk 4 above)
Comment (kobolog) [inline include/cocaine/detail/engine.hpp]: So it's because session is now a template, not because conversion. Okay.

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (kobolog) [inline debian/cocaine-v12-runtime.init]: If it's a Yandex-related problem, solve it internally, don't put this bullshit here.

File: include/cocaine/rpc/session.hpp (hunk 7 above)
Comment (3Hren) [inline include/cocaine/rpc/session.hpp]: There is a TODO several lines above :)

AFAIK currently it is required by encoder (which owns the header table) for its purposes related with header table.

Comment (kobolog) [inline include/cocaine/rpc/session.hpp]: So I'll cut it out together with headers for now.

File: debian/cocaine-v12-runtime.init (hunk 3 above)
Comment (oktocat) [inline debian/cocaine-v12-runtime.init]: 1. Workflows were introduced in conductor for this exact reason - to handle non-standart package isnt in all repos and/or should be dmoved differently in different repos. I suggest you to try to use that stuff, even it's unclear and vague in a way.
2.  Have you addressed Jay on this problem? Because this kind of "workaround" looks to me as an ugly-dirty-fucking-hack.
¯\_(ツ)_/¯

Comment (oktocat) [inline debian/cocaine-v12-runtime.init]: And oh, it reminds me of a story with elliptics and elliptics-2.10(elliptics-2.10=2.23.1.2.3, lol). Go to the next room and ask Andrew Godin about the PAIN if you don't believe me.

The easiest and not so dirty solution is to distribute apt pinning settings together with cocaine-runtime v0.11 package.
